runtime-light/stdlib/array/array-functions.h

```diff
@@ -5,16 +5,218 @@
 #pragma once
 
 #include <concepts>
+#include <cstdint>
 #include <functional>
 #include <utility>
 
 #include "runtime-common/core/runtime-core.h"
+#include "runtime-common/stdlib/array/array-functions.h"
 #include "runtime-light/coroutine/task.h"
 #include "runtime-light/stdlib/math/random-functions.h"
 #include "runtime-light/utils/concepts.h"
 
+namespace dl {
+
+template<typename T, typename Comparator>
+requires(std::invocable<Comparator, T, T>) task_t<void> sort(T *begin_init, T *end_init, Comparator compare) noexcept {
+  auto compare_call = [compare]<typename U>(U lhs, U rhs) -> task_t<int64_t> {
+    if constexpr (is_async_function_v<Comparator, U, U>) {
+      co_return co_await std::invoke(std::move(compare), std::move(lhs), std::move(rhs));
+    } else {
+      co_return std::invoke(std::move(compare), std::move(lhs), std::move(rhs));
+    }
+  };
+  T *begin_stack[32];
+  T *end_stack[32];
+
+  begin_stack[0] = begin_init;
+  end_stack[0] = end_init - 1;
+
+  for (int depth = 0; depth >= 0; --depth) {
+    T *begin = begin_stack[depth];
+    T *end = end_stack[depth];
+
+    while (begin < end) {
+      const auto offset = (end - begin) >> 1;
+      swap(*begin, begin[offset]);
+
+      T *i = begin + 1;
+      T *j = end;
+
+      while (true) {
+        while (i < j && (co_await compare_call(*begin, *i)) > 0) {
```

Comment:
We can get rid of `co_await` completely unless `Comparator` is async function.

Reply:
Yes, we can, but it require to create copy paste function without co_await. Is it good decision - I don't think so. In any case, we should make base sort functions not interruptible, it's part of a larger task

Reply:
fix